project/visitorckw/README.md

```diff
@@ -0,0 +1,228 @@
+of each element to enable correct sorting. The provided array can be of any data type.
+- Linked List: For linked list sorting, a C++ class follows Linux kernel-style coding
+conventions to encapsulate the linked list structure, making it convenient and user-friendly.
+
+### Sorting Methods:
+
+The library offers a selection of sorting methods, which are defined using an enum called
+SortingMethod. These methods include:
+- QUICKSORT
+- MERGESORT
+- HEAPSORT
+- BINARY_INSERTION_SORT
+- TIMSORT
+- INTROSORT
+
+### Functionality:
+
+- Recommend Sorting Method: The library provides a function that takes test data and a
+memory limit as input and recommends the most suitable sorting method based on the
+characteristics of the data and memory constraints.
+- Sorting Functions: Users can invoke sorting functions for both arrays and linked lists,
+specifying the sorting method they wish to use.
+- Performance Measurement: The library includes functions to measure the execution time,
+comparison counts, and swap counts for a given sorting task.
+- Result Visualization: A function is available to create graphs that visualize the results
+of performance tests.
+
+
+This system architecture allows for easy integration of various sorting algorithms and
+provides a user-friendly interface for developers working with different data structures.
+The ability to recommend sorting methods based on data characteristics enhances the
+adaptability of the library, ensuring that users can choose the most efficient sorting
+approach for their specific use cases. The modular structure of the library facilitates
+future extensions and the addition of new sorting methods or optimizations as needed.
+
+## API Description
+C++ API:
+* `enum SortingMethod { QUICKSORT, MERGESORT, HEAPSORT, BINARY_INSERTION_SORT, TIMSORT, INTROSORT };`
+* `SortingMethod RecommendSortingMethod(const std::vector<T>& data, size_t memoryLimit);`
+* `void SortArray(std::vector<int>& data, SortingMethod method);`
+* `void SortLinkedList(struct list_head* head, SortingMethod method);`
+* `void MeasureSortingPerformance(const std::vector<T>& data, SortingMethod method, double& executionTime, size_t& comparisons, size_t& swaps);`
+* `void PlotTestResults(const std::vector<double>& executionTimes, const std::vector<size_t>& comparisons, const std::vector<size_t>& swaps);`
+
+## Engineering Infrastructure
+* Automatic build system: `make`
+* Version control: `git`
+* Testing framework: `pytest`
+* Documentation: GitHub `README.md`
+
+
+## Schedule
+
+Planning phase (6 weeks from 9/18 to 10/30): Think up presentation, then set up the environment
+and get familiar with the algorithm
+
+Week 1 (10/30):  Project Initiation
+
+- Define project objectives, scope, and requirements.
+- Set up the development environment and tools.
+- Create a project plan, including milestones and deliverables.
+
+Week 2 (11/6): Data Structures and API Design
+
+- Design the data structures for arrays and linked lists.
```

Comment:
Do you have a plan for memory management?  Array and linked list use memory in different ways.